libgraphics/src/graphics-internal.h

```diff
@@ -36,43 +36,227 @@ along with LiveCode.  If not see <http://www.gnu.org/licenses/>.  */
 
 ////////////////////////////////////////////////////////////////////////////////
 
-typedef struct __MCGPattern *MCGPatternRef;
+typedef class MCGObject *MCGObjectRef;
 
-struct __MCGPattern
+class MCGObject
 {
-	MCGImageRef			pattern;
-	MCGAffineTransform	transform;
-	MCGImageFilter		filter;
-	uint32_t			references;
+public:
+#ifdef _DEBUG
+    static size_t s_object_count;
+#endif
+
+    MCGObject(void)
+        : m_references(1)
+    {
+#ifdef _DEBUG
+        s_object_count += 1;
+#endif
+    }
+
+protected:
+    virtual ~MCGObject(void)
+    {
+#ifdef _DEBUG
+        s_object_count -= 1;
+#endif
+    }
+
+private:
+    virtual void Retain(void)
+    {
+        m_references += 1;
+    }
+    
+    virtual void Release(void)
+    {
+        m_references -= 1;
+        if (m_references == 0)
+        {
+            delete this;
+        }
+    }
+    
+    template<typename T>
+    friend T MCGRetain(T p_object)
+    {
+        if (p_object == nullptr)
+        {
+            return nullptr;
+        }
+        p_object->Retain();
+        return p_object;
+    }
+    
+    template<typename T>
+    friend void MCGRelease(T p_object)
+    {
+        if (p_object == nullptr)
+        {
+            return;
+        }
+        p_object->Release();
+    }
+    
+    template<typename T>
+    friend void MCGAssignAndRelease(T& x_target, T p_source)
+    {
+        if (p_source == x_target)
+        {
+            return;
```

Comment:
Since `p_source` comes in here with >0 ref count, don't we need to release it to prevent memory leaking?

Reply:
It is 'AssignAndRelease' - so x_target takes the ref count from p_source... i.e. We release x_target, but *not* Retain p_source.